### DISCUSSION TITLE
Trivy not detecting wildcards in policy

### DISCUSSION THREAD

**Discussion by hirra-farooq** [aquasecurity/trivy#7878] [Bugs]
I am migrating from tfsec to trivy. I have noticed that there are some issues that trivy is not detecting that tfsec usually would. 
One example that i think is especially important, given the security principles of least privilege is:
```
data "aws_iam_policy_document" "test_policy" {
  statement {
    sid    = "Allowall"
    effect = "Allow"

    actions = ["s3*"]

    resources = ["*"]
  }
}
```
tfsec would raise:
https://aquasecurity.github.io/tfsec/v1.28.6/checks/aws/iam/no-policy-wildcards/
- HIGH IAM policy document uses wildcarded action 's3*' 
- HIGH IAM policy document uses sensitive action 's3*' on wildcarded resource '*' 

But doing:
`trivy config ./myfolder` does not raise any issues. 

I would have thought the trivy misconfiguration: https://avd.aquasec.com/misconfig/aws/iam/avd-aws-0057/ would be raised? 
Is this behaviour intended or a bug?

**Comment by nikpivkin** [comment]
Hi @hirra-farooq !

Check `AVD-AWS-0057` is now [deprecated](https://github.com/aquasecurity/trivy-checks/blob/71746ccff203a455296210c30823a6235b36faa7/checks/cloud/aws/iam/no_policy_wildcards.go#L53).

**Reply by ponkio-o** [reply]
@nikpivkin Hi,
I understand that it is not a bug, but why was it deprecated? Please let me know if there is an Issue where the background was discussed

**Reply by ken5scal** [reply]
Hi @nikpivkin 
let me add other questions.

- trivy 0.58.0 with `--include-deprecated-checks` option didn't detect the same config. Is this behavior intended or a bug? or does it only work with "Rego" rules
- This page doesn't show the status of the rule. Is it safe to assume the page does not show the status or it's just preparing to update the status. https://avd.aquasec.com/misconfig/aws/iam/avd-aws-0057/

Thanks in advance!

```
% trivy version
Version: 0.58.0
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-12-11 12:17:09.376381063 +0000 UTC
  NextUpdate: 2024-12-12 12:17:09.376380663 +0000 UTC
  DownloadedAt: 2024-12-11 14:18:47.690554 +0000 UTC
Check Bundle:
  Digest: sha256:f6901e03f486a48f47aa17a78d89d18e6c31ded82aff83ed19d0d73935a1a059
  DownloadedAt: 2024-12-12 01:00:41.712391 +0000 UTC
% trivy config --include-deprecated-checks .
2024-12-12T10:41:42+09:00       INFO    [misconfig] Misconfiguration scanning is enabled
2024-12-12T10:41:47+09:00       INFO    [terraform scanner] Scanning root module        file_path="."
2024-12-12T10:41:47+09:00       INFO    [terraform scanner] Scanning root module        file_path="tmp"
2024-12-12T10:41:48+09:00       INFO    Detected config files   num=2
```

**Reply by ken5scal** [reply]
Adding more information, it seems trivy 0.56.0 with `--include-deprecated-checks`  detect "avd-aws-0057"

```
% docker run --rm -v $PWD:/cfn aquasec/trivy:0.56.0 conf /cfn --include-deprecated-checks

Unable to find image 'aquasec/trivy:0.56.0' locally
0.56.0: Pulling from aquasec/trivy
cf04c63912e1: Already exists 
e0bb7b1f10c6: Pull complete 
8ae5d7faa388: Pull complete 
61648e7cdbb8: Pull complete 
Digest: sha256:e682a9f8db9db6f09731fe2f63e4234d240ec67a811d90974b5af3d40497f843
Status: Downloaded newer image for aquasec/trivy:0.56.0
2024-12-12T01:49:07Z    INFO    [misconfig] Misconfiguration scanning is enabled
2024-12-12T01:49:07Z    INFO    [misconfig] Need to update the built-in checks
2024-12-12T01:49:07Z    INFO    [misconfig] Downloading the built-in checks...
160.80 KiB / 160.80 KiB [------------------------------------------------------] 100.00% ? p/s 200ms2024-12-12T01:49:15Z        INFO    [terraform scanner] Scanning root module    file_path="."
2024-12-12T01:49:15Z    INFO    Detected config files   num=2

sample_resource.tf (terraform)
==============================
Tests: 2 (SUCCESSES: 0, FAILURES: 2, EXCEPTIONS: 0)
Failures: 2 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 2, CRITICAL: 0)

HIGH: IAM policy document uses sensitive action '*' on wildcarded resource '*'
════════════════════════════════════════
You should use the principle of least privilege when defining your IAM policies. This means you should specify each exact permission required without using wildcards, as this could cause the granting of access to certain undesired actions, resources and principals.

See https://avd.aquasec.com/misconfig/avd-aws-0057
────────────────────────────────────────
 sample_resource.tf:33
   via sample_resource.tf:33 (aws_iam_role_policy.sample.policy)
    via sample_resource.tf:32-41 (aws_iam_role_policy.sample.policy)
     via sample_resource.tf:29-42 (aws_iam_role_policy.sample)
────────────────────────────────────────
  29   resource "aws_iam_role_policy" "sample" {
  ..   
  33 [     Version = "2012-10-17",
  ..   
  42   }
────────────────────────────────────────


HIGH: IAM policy document uses wildcarded action '*'
════════════════════════════════════════
You should use the principle of least privilege when defining your IAM policies. This means you should specify each exact permission required without using wildcards, as this could cause the granting of access to certain undesired actions, resources and principals.

See https://avd.aquasec.com/misconfig/avd-aws-0057
────────────────────────────────────────
 sample_resource.tf:33
   via sample_resource.tf:33 (aws_iam_role_policy.sample.policy)
    via sample_resource.tf:32-41 (aws_iam_role_policy.sample.policy)
     via sample_resource.tf:29-42 (aws_iam_role_policy.sample)
────────────────────────────────────────
  29   resource "aws_iam_role_policy" "sample" {
  ..   
  33 [     Version = "2012-10-17",
  ..   
  42   }
────────────────────────────────────────
```

**Reply by knqyf263** [reply]
Some context here. 
https://github.com/aquasecurity/trivy-checks/pull/245

@nikpivkin We should improve AVD to display information about deprecations.
https://avd.aquasec.com/misconfig/aws/iam/avd-aws-0057/

**Reply by nikpivkin** [reply]
@itaysk We can add information that the check is deprecated to the header in the check list. For example:
<img width="766" alt="image" src="https://github.com/user-attachments/assets/8c03714d-42c6-4285-8777-5af9382790b4" />

And also in the table with the check metadata:
<img width="355" alt="image" src="https://github.com/user-attachments/assets/83706deb-33b1-4a3a-8eb0-11c0c2ec111f" />
WDYT?

**Reply by knqyf263** [reply]
LGTM. Let's start with that. After Simar is back, we can discuss it again.

**Reply by simar7** [reply]
@nikpivkin @knqyf263 the above UI change lgtm. I think we can also add reasoning to the AVD page as to why it was deprecated.

**Comment by markdboyd** [comment]
I can't seem to get https://avd.aquasec.com/misconfig/avd-aws-0057 to detect on bad configuration even when specifying `--include-deprecated-checks` or including this in `trivy.yaml`:

```
rego:
  # Same as '--include-deprecated-checks'
  include-deprecated-checks: true
```

**Comment by visit1985** [comment]
The deprecation reason doesn't seem reasonable to me.

>These checks used data extracted from HTML pages to obtain information about allowed wildcards in actions. This data should be updated regularly to avoid false positives. Since AWS does not provide an official API to obtain a list of wildcard-enabled resources, this approach cannot be considered reliable.

People are relying on trivy to find those wildcards and don't even recognize that it stopped scanning for them.

Is there any update on this?

**Reply by simar7** [reply]
Trivy does not want to report false positives since AWS does not provide an official API to be able to obtain a list of wildcard-enabled resources. Maintaining such a list ourselves is simply not feasible at this time. Hence the depreciation.

**Reply by visit1985** [reply]
I understand your goal to eliminate false positives and that it is not ideal to maintain such a list. But I would rather opt for community support to maintain such a list instead of not checking policies at all.

Trivy states in its docs to provide CIS AWS Foundations Benchmark compliance reports, which is essentially not true because of incomplete coverage.

As @markdboyd stated above we can't even workaround this, because `include-deprecated-checks` does not work anymore.

**Reply by simar7** [reply]
> Trivy states in its docs to provide CIS AWS Foundations Benchmark compliance reports, which is essentially not true because of incomplete coverage.

Full set of compliance capability is available in the commercial version as mentioned here https://trivy.dev/latest/commercial/compare/#iaccspm-scanning

> As @markdboyd stated above we can't even workaround this, because include-deprecated-checks does not work anymore.

This doesn't seem to be related to us deprecating the check itself but sounds like a bug to me. @nikpivkin could you take a look?

**Reply by nikpivkin** [reply]
> This doesn't seem to be related to us deprecating the check itself but sounds like a bug to me. @nikpivkin could you take a look?

We have removed all Go checks, and some wildcard related checks in IAM policies have not been migrated to Rego, so the `include-deprecated-checks` flag will not include them.
